### PR TITLE
Add signers to Options class

### PR DESCRIPTION
**Pull request by juliangums**

### CODE REVIEW

File: src/Payload/Options.php [hunk 1]
```diff
@@ -11,6 +11,7 @@ public function __construct(
         public readonly ?bool $immutable = null,
         public readonly ?bool $forceAccount = null,
         public readonly ?ReturnUrl $returnUrl = null,
+        public readonly ?array $signers = null,
```
Comment (paulinevos) [inline src/Payload/Options.php:14]: Hi! Thanks for the PR. Could you make this `string ...$signers` instead or guarantee string type in some other way? 

And testing that this is properly serialized would be nice as well. Thanks in advance.

**Review by paulinevos — CHANGES_REQUESTED** [review]

**Comment by Dennis-Mwea** [comment]
@paulinevos Is there a restriction for not using PHPDoc for type hinting?

**Comment by juliangums** [comment]
Unfortunately, you can't declare an array of strings directly. 

We could add a docblocks like below but this would not enforce anything, only suggest it
```
    /**
     * @param array<string>|null $signers
     */
```

Then you could create a class that extends ArrayObject and add some type hinting and validation. Is there maybe some open-source lib that helps with this? I couldn't find anything.

The only solution without creating a new class is adding a method to check and a docblock to hint:
```php
<?php
declare (strict_types=1);
namespace Xrpl\XummSdkPhp\Payload;

final class Options
{
    /**
     * @param array<string>|null $signers
     */
    public function __construct(
        public readonly ?bool $submit = null,
        public readonly ?bool $multisign = null,
        public readonly ?int $expire = null,
        public readonly ?bool $immutable = null,
        public readonly ?bool $forceAccount = null,
        public readonly ?array $signers = null,
        public readonly ?ReturnUrl $returnUrl = null,
    ) {
        if($signers) $this->typeArrayNullString(...$signers);
    }
    private function typeArrayNullString(?string ...$arg): void {
    }
}
```

Let me know if that's ok @paulinevos .

**Comment by paulinevos** [comment]
Hi @juliangums, we can enforce it through the `string ...$signers` argument notation as per the suggestion in the comment.

**Comment by Dennis-Mwea** [comment]
```
final class Options
{
     public readonly ?array $signers;

    public function __construct(
        public readonly ?bool $submit = null,
        public readonly ?bool $multisign = null,
        public readonly ?int $expire = null,
        public readonly ?bool $immutable = null,
        public readonly ?bool $forceAccount = null,
        ?string ...$signers,
    )
    {
        $this->signers = $signers;
    }
}
```

**Comment by paulinevos** [comment]
@Dennis-Mwea exactly like that, although you could even remove the `?` notation. It would simply be assigned to an empty array if I'm not mistaken. Good to double-check with a test case though!

**Comment by Dennis-Mwea** [comment]
The limitation with this approach is you'd have to pass all params `new Options(null, null, null, null, null, 'er4tyui', 'qwertyui')` for you to be able to pass multiple signers. at the end. Also, you can't add more parameters at the end of the initialiser because they will be interpreted as part of the signers array.

**Comment by paulinevos** [comment]
@Dennis-Mwea That's definitely true, which is why the docs suggest using named arguments for constructing payloads, as most of it is nullable. Then you won't have to pass `null` for everything else. It's either this or validating type some other way, imo this is the most elegant but some other way is fine too. Most important thing is to enforce type safety in some way :)

**Comment by juliangums** [comment]
@paulinevos like so? `$options = new Options(signers:["sdfsdf"]);`
That fails on my end as it is expecting a string
`TypeError: Xrpl\XummSdkPhp\Payload\Options::__construct(): Argument #6 must be of type ?string, array given on line 1`

**Comment by paulinevos** [comment]
@juliangums Good point, I think in this case we should prefer to be able to use named arguments. So I'd suggest passing it the way you passed it before and adding a check that all the array elements are strings. i.e., throw an exception in the constructor if they're not all strings. Then add a unit test for the `Options` constructor.

File: src/Payload/Options.php [hunk 2]
```diff
@@ -1,17 +1,23 @@
...
         public readonly ?int $expire = null,
         public readonly ?bool $immutable = null,
         public readonly ?bool $forceAccount = null,
-        public readonly ?ReturnUrl $returnUrl = null,
         public readonly ?array $signers = null,
+        public readonly ?ReturnUrl $returnUrl = null,
     ) {
+        if($signers) $this->typeArrayNullString(...$signers);
+    }
+    private function typeArrayNullString(?string ...$arg): void {
```
Comment (paulinevos) [inline src/Payload/Options.php:21]: Thanks for the update. Could you please throw an exception here? Refer to the `Exception` namespace to see how other exceptions are created. Maybe something like an `InvalidOptionsException` within the `Exception\Payload` namespace.

And again a test case would be good. Ideally a unit test for the constructor asserting the exception is thrown, and to add some `signers` strings to the serialization test

Comment (juliangums) [reply src/Payload/Options.php]: @paulinevos it throws an exception as I’m using `declare (strict_types=1);`. Thought that looks a bit neater. @Dennis-Mwea also added the unit test. Good to go?

**Comment by paulinevos** [comment]
Thanks @juliangums and @Dennis-Mwea. I do have a strong preference for exceptions that can be handled properly by the user. For instance, users of the package can listen for specific `XummLogicException`s and log them or act upon them specifically. That's why all the exceptions in the package implement `XummLogicException`. 

Apart from that it looks great! There's some small issues being picked up by the linter that is failing checks. Those can be fixed automatically by running `make fix-cs` or `./vendor/bin/phpcbf --standard=PSR12 ./src`

**Comment by Dennis-Mwea** [comment]
On it.

**Comment by paulinevos** [comment]
Awesome 🙏 thanks for your persistence

**Comment by Dennis-Mwea** [comment]
@paulinevos, @juliangums thoughts?
```
final class Options
{
    /**
     * @throws InvalidArgumentException
     */
    public function __construct(
        public readonly ?bool $submit = null,
        public readonly ?bool $multisign = null,
        public readonly ?int $expire = null,
        public readonly ?bool $immutable = null,
        public readonly ?bool $forceAccount = null,
        public readonly ?ReturnUrl $returnUrl = null,
        public readonly ?array $signers = null,
    ) {
        foreach ($this->signers as $key => $signer) {
            $this->validate($signer, $key);
        }
    }

    /**
     * @param $value
     * @param int $key
     * @return void
     * @throws InvalidArgumentException
     */
    private function validate($value, int $key): void
    {
        if (!is_string($value)) {
            $type = gettype($value);
            throw new InvalidArgumentException("Argument #$key must be of type ?string, $type given");
        }
    }
}```

**Comment by paulinevos** [comment]
That looks excellent :)

**Comment by paulinevos** [comment]
@juliangums I would be but I think you accidentally removed the `signers` constructor parameter/property. So if you add that back, then yes :) Don't forget to run tests before the push to ensure checks are passing.

**Comment by juliangums** [comment]
Done

**Comment by paulinevos** [comment]
@juliangums Did you run `make test`? There's one failing acceptance test

**Comment by paulinevos** [comment]
The test is failing because `null` can be passed and you're then attempting to loop over `null`, which throws a warning. 
Wrapping the `foreach` in a `if ($this->signers !== null)` will do the trick.

**Review by paulinevos — APPROVED** [review]

**Comment by paulinevos** [comment]
Thanks for your contribution, merging now! I'll do a release today or over the weekend that will include your changes.

**Pull request merged**